File: internal/forms/solver.go
package forms

import (
	"fmt"
	"slices"
	"strings"
)

// DependencyGraph builds and resolves the DAG of all form fields.
type DependencyGraph struct {
	registry *Registry
	// adjacency list: key -> list of keys it depends on
	edges map[string][]string
	// all known field keys
	nodes map[string]bool
}

// NewDependencyGraph creates a new DependencyGraph backed by the given registry.
func NewDependencyGraph(registry *Registry) *DependencyGraph {
	return &DependencyGraph{
		registry: registry,
		edges:    make(map[string][]string),
		nodes:    make(map[string]bool),
	}
}

// Build constructs the dependency graph from all registered forms.
// Returns an error if a cycle is detected.
func (g *DependencyGraph) Build() error {
	g.edges = make(map[string][]string)
	g.nodes = make(map[string]bool)

	for _, form := range g.registry.AllForms() {
		for _, field := range form.Fields {
			key := FieldKey(form.ID, field.Line)
			g.nodes[key] = true

			// Resolve dependencies. Wildcard deps expand to all matching keys.
			var resolved []string
			for _, dep := range field.DependsOn {
				if strings.Contains(dep, "*") {
					// Expand wildcard to all matching registered field keys
					for node := range g.nodes {
						if matchWildcard(dep, node) {
							resolved = append(resolved, node)
						}
					}
					// Also check nodes added later — we do a second pass below
				} else {
					resolved = append(resolved, dep)
				}
			}
			g.edges[key] = resolved
		}
	}

	// Second pass: expand wildcards again now that all nodes are known.
	for _, form := range g.registry.AllForms() {
		for _, field := range form.Fields {
			key := FieldKey(form.ID, field.Line)
			for _, dep := range field.DependsOn {
				if strings.Contains(dep, "*") {
					existing := make(map[string]bool)
					for _, e := range g.edges[key] {
						existing[e] = true
					}
					for node := range g.nodes {
						if matchWildcard(dep, node) && !existing[node] {
							g.edges[key] = append(g.edges[key], node)
						}
					}
				}
			}
		}
	}

	// Check for cycles
	_, err := g.TopologicalSort()
	return err
}

// TopologicalSort returns all field keys in dependency order using Kahn's algorithm.
// Fields with no dependencies come first. Returns an error if a cycle is detected.
func (g *DependencyGraph) TopologicalSort() ([]string, error) {
	// Compute in-degree for each node
	inDegree := make(map[string]int)
	for node := range g.nodes {
		inDegree[node] = 0
	}

	// Build reverse adjacency: for each edge (key depends on dep),
	// dep -> key means dep must come before key
	reverseAdj := make(map[string][]string)
	for key, deps := range g.edges {
		for _, dep := range deps {
			// Only count edges to known nodes
			if g.nodes[dep] {
				inDegree[key]++
				reverseAdj[dep] = append(reverseAdj[dep], key)
			}
		}
	}

	// Start with nodes that have zero in-degree
	var queue []string
	for node := range g.nodes {
		if inDegree[node] == 0 {
			queue = append(queue, node)
		}
	}

	// Sort queue for deterministic output
	slices.Sort(queue)

	var sorted []string
	for len(queue) > 0 {
		node := queue[0]
		queue = queue[1:]
		sorted = append(sorted, node)

		for _, dependent := range reverseAdj[node] {
			inDegree[dependent]--
			if inDegree[dependent] == 0 {
				queue = append(queue, dependent)
				slices.Sort(queue)
			}
		}
	}

	if len(sorted) != len(g.nodes) {
		// Find nodes involved in cycle for error message
		var cycleNodes []string
		for node, deg := range inDegree {
			if deg > 0 {
				cycleNodes = append(cycleNodes, node)
			}
		}
		slices.Sort(cycleNodes)
		return nil, fmt.Errorf("cycle detected involving fields: %s", strings.Join(cycleNodes, ", "))
	}

	return sorted, nil
}

// MissingInputs returns all UserInput field keys that don't have values in the
// provided map.
func (g *DependencyGraph) MissingInputs(provided map[string]float64) []string {
	var missing []string
	for _, form := range g.registry.AllForms() {
		for _, field := range form.Fields {
			if field.Type == UserInput {
				key := FieldKey(form.ID, field.Line)
				if _, ok := provided[key]; !ok {
					missing = append(missing, key)
				}
			}
		}
	}
	slices.Sort(missing)
	return missing
}

// Solve resolves all Computed fields given the provided UserInput values.
// Returns all field values (inputs + computed). Returns an error if required
// UserInput values are missing.
func (g *DependencyGraph) Solve(inputs map[string]float64, strInputs map[string]string, taxYear int) (map[string]float64, error) {
	// Check for missing required inputs
	missing := g.MissingInputs(inputs)
	if len(missing) > 0 {
		return nil, fmt.Errorf("missing required inputs: %s", strings.Join(missing, ", "))
	}

	order, err := g.TopologicalSort()
	if err != nil {
		return nil, err
	}

	// Initialize result with provided inputs
	result := make(map[string]float64)
	for k, v := range inputs {
		result[k] = v
	}

	strResult := make(map[string]string)
	for k, v := range strInputs {
		strResult[k] = v
	}

	// Process fields in topological order
	for _, key := range order {
		if _, hasValue := result[key]; hasValue {
			continue // already provided as input
		}

		_, field, err := g.registry.GetField(key)
		if err != nil {
			continue
		}

		switch field.Type {
		case UserInput:
			// Should already be in result; if not it's an error caught above
			continue
		case Computed, Lookup, FederalRef, PriorYear:
			dv := NewDepValues(result, strResult, taxYear)
			if field.Compute != nil {
				result[key] = field.Compute(dv)
			}
			if field.ComputeStr != nil {
				strResult[key] = field.ComputeStr(dv)
			}
		}
	}

	return result, nil
}

File: api/src/domain/solver.rs
use std::collections::{BTreeSet, HashMap, HashSet};

use crate::domain::field::{match_wildcard, DepValues, FieldType};
use crate::domain::registry::Registry;

/// DependencyGraph builds and resolves the DAG of all form fields.
pub struct DependencyGraph<'a> {
    registry: &'a Registry,
    /// Adjacency list: key -> list of keys it depends on.
    edges: HashMap<String, Vec<String>>,
    /// All known field keys.
    nodes: HashSet<String>,
}

impl<'a> DependencyGraph<'a> {
    /// Creates a new DependencyGraph backed by the given registry.
    pub fn new(registry: &'a Registry) -> Self {
        Self {
            registry,
            edges: HashMap::new(),
            nodes: HashSet::new(),
        }
    }

    /// Constructs the dependency graph from all registered forms.
    /// Returns an error if a cycle is detected.
    pub fn build(&mut self) -> Result<(), String> {
        self.edges.clear();
        self.nodes.clear();

        // First pass: register all nodes and expand wildcards against known nodes so far
        for form in self.registry.all_forms() {
            for field in &form.fields {
                let key = format!("{}:{}", form.id, field.line);
                self.nodes.insert(key.clone());

                let mut resolved = Vec::new();
                for dep in &field.depends_on {
                    if dep.contains('*') {
                        // Expand wildcard to all matching registered field keys
                        for node in &self.nodes {
                            if match_wildcard(dep, node) {
                                resolved.push(node.clone());
                            }
                        }
                    } else {
                        resolved.push(dep.clone());
                    }
                }
                self.edges.insert(key, resolved);
            }
        }

        // Second pass: expand wildcards again now that all nodes are known
        for form in self.registry.all_forms() {
            for field in &form.fields {
                let key = format!("{}:{}", form.id, field.line);
                for dep in &field.depends_on {
                    if dep.contains('*') {
                        let existing: HashSet<String> = self
                            .edges
                            .get(&key)
                            .map(|v| v.iter().cloned().collect())
                            .unwrap_or_default();

                        let mut new_deps = Vec::new();
                        for node in &self.nodes {
                            if match_wildcard(dep, node) && !existing.contains(node) {
                                new_deps.push(node.clone());
                            }
                        }
                        if let Some(edge_list) = self.edges.get_mut(&key) {
                            edge_list.extend(new_deps);
                        }
                    }
                }
            }
        }

        // Check for cycles via topological sort
        self.topological_sort()?;
        Ok(())
    }

    /// Returns all field keys in dependency order using Kahn's algorithm.
    /// Fields with no dependencies come first.
    pub fn topological_sort(&self) -> Result<Vec<String>, String> {
        // Compute in-degree for each node
        let mut in_degree: HashMap<&str, usize> = HashMap::new();
        for node in &self.nodes {
            in_degree.insert(node.as_str(), 0);
        }

        // Build reverse adjacency: dep -> list of keys that depend on it
        let mut reverse_adj: HashMap<&str, Vec<&str>> = HashMap::new();
        for (key, deps) in &self.edges {
            for dep in deps {
                // Only count edges to known nodes
                if self.nodes.contains(dep) {
                    *in_degree.entry(key.as_str()).or_insert(0) += 1;
                    reverse_adj
                        .entry(dep.as_str())
                        .or_default()
                        .push(key.as_str());
                }
            }
        }

        // Start with nodes that have zero in-degree (sorted for determinism)
        let mut queue: BTreeSet<&str> = BTreeSet::new();
        for (node, &deg) in &in_degree {
            if deg == 0 {
                queue.insert(node);
            }
        }

        let mut sorted = Vec::new();
        while let Some(node) = queue.iter().next().copied() {
            queue.remove(node);
            sorted.push(node.to_string());

            if let Some(dependents) = reverse_adj.get(node) {
                for &dependent in dependents {
                    if let Some(deg) = in_degree.get_mut(dependent) {
                        *deg -= 1;
                        if *deg == 0 {
                            queue.insert(dependent);
                        }
                    }
                }
            }
        }

        if sorted.len() != self.nodes.len() {
            // Find nodes involved in cycle for error message
            let mut cycle_nodes: Vec<&str> = in_degree
                .iter()
                .filter(|&(_, deg)| *deg > 0)
                .map(|(&node, _)| node)
                .collect();
            cycle_nodes.sort();
            return Err(format!(
                "cycle detected involving fields: {}",
                cycle_nodes.join(", ")
            ));
        }

        Ok(sorted)
    }

    /// Returns all UserInput field keys that don't have values in the provided map.
    pub fn missing_inputs(&self, provided: &HashMap<String, f64>) -> Vec<String> {
        let mut missing = Vec::new();
        for form in self.registry.all_forms() {
            for field in &form.fields {
                if field.field_type == FieldType::UserInput {
                    let key = format!("{}:{}", form.id, field.line);
                    if !provided.contains_key(&key) {
                        missing.push(key);
                    }
                }
            }
        }
        missing.sort();
        missing
    }

    /// Resolves all Computed fields given the provided UserInput values.
    /// Returns all field values (inputs + computed).
    pub fn solve(
        &self,
        inputs: &HashMap<String, f64>,
        str_inputs: &HashMap<String, String>,
        tax_year: i32,
    ) -> Result<HashMap<String, f64>, String> {
        // Check for missing required inputs
        let missing = self.missing_inputs(inputs);
        if !missing.is_empty() {
            return Err(format!(
                "missing required inputs: {}",
                missing.join(", ")
            ));
        }

        let order = self.topological_sort()?;

        // Initialize result with provided inputs
        let mut result: HashMap<String, f64> = inputs.clone();
        let mut str_result: HashMap<String, String> = str_inputs.clone();

        // Process fields in topological order
        for key in &order {
            if result.contains_key(key) {
                continue; // already provided as input
            }

            let (_form, field) = match self.registry.get_field(key) {
                Ok(v) => v,
                Err(_) => continue,
            };

            match field.field_type {
                FieldType::UserInput => {
                    // Should already be in result
                    continue;
                }
                FieldType::Computed
                | FieldType::Lookup
                | FieldType::FederalRef
                | FieldType::PriorYear => {
                    let dv = DepValues::new(result.clone(), str_result.clone(), tax_year);
                    if let Some(ref compute) = field.compute {
                        result.insert(key.clone(), compute(&dv));
                    }
                    if let Some(ref compute_str) = field.compute_str {
                        str_result.insert(key.clone(), compute_str(&dv));
                    }
                }
            }
        }

        Ok(result)
    }
}
